src/Microsoft.OData.Client/BulkUpdateGraph.cs

```diff
@@ -0,0 +1,141 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="BulkUpdateGraph.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.OData.Client
+{
+    /// <summary>
+    /// Contains information about descriptors and their related descriptors. 
+    /// </summary>
+    internal sealed class BulkUpdateGraph
+    {
+        /// <summary>Set of related descriptors</summary>
+        private Dictionary<Descriptor, HashSet<Descriptor>> descriptorGraph;
+
+        /// <summary>Set of top level descriptors</summary>
+        private List<EntityDescriptor> topLevelDescriptors;
+
+        /// <summary> The EntitySetName of the top-level descriptor types </summary>
+        private string entitySetName;
+
+        /// <summary>where to pull the changes from</summary>
+        protected readonly RequestInfo requestInfo;
+
+        public BulkUpdateGraph(RequestInfo requestInfo)
+        { 
+            this.descriptorGraph= new Dictionary<Descriptor, HashSet<Descriptor>>(); 
+            this.topLevelDescriptors= new List<EntityDescriptor>();
+            this.requestInfo = requestInfo;
+        }
+
+        /// <summary>
+        /// Returns a collection of all the root descriptors used in a deep update.
+        /// </summary>
+        public List<EntityDescriptor> TopLevelDescriptors
+        {
+            get
+            {
+                return this.topLevelDescriptors;
+            }
+        }
+
+        /// <summary>
+        /// Returns the EntitySetName.
+        /// </summary>
+        public string EntitySetName
+        {
+            get
+            {
+                return this.entitySetName;
+            }
+        }
+
+        /// <summary>
+        /// Computes the EntyitySetName in cases where the top-level EntityDescriptors
+        /// do not have the EntitySetName set. 
+        /// </summary>
+        public void ComputeEntitySetName()
+        {
+            EntityDescriptor parentDescriptor = this.TopLevelDescriptors[0] as EntityDescriptor;

```

Comment:
Have made this method `internal`. Also, the method gets called after checking the `TopLevelDescriptors` and ensuring that it is not empty.

Reply:
I don't think you get my argument... Remember quite a few people collaborate on this project. If someone is unfamiliar with how the method is implemented, how will they know that they should check that `TopLevelDescriptors` collection is not empty before calling the method? The method should "protect" itself from possible out-of-range exception without depending on what the caller might or might not do...

Reply:
I think I made the changed you are suggesting on this methid. My comment was before I made those changes.

Reply:
It's even worse now since the method is triggered by accessing a property.
Too much back and forth... What do you have against doing this:
```csharp
if (this.TopLevelDescriptions.Count == 0)
{
    return;
}
```

Reply:
Have added the check.